[excel, CommonAPI] (Code patterns) Adding notes about range limits and promiseless callback methods (#1406)
OfficeDev/office-js-docs-pr: docs/develop/common-coding-issues.md

---
title: Common coding issues and unexpected platform behaviors
description: 'A list of Office JavaScript API platform issues frequently encountered by developers.'
ms.date: 10/30/2019
localization_priority: Normal
---

# Common coding issues and unexpected platform behaviors

This article highlights aspects of the Office JavaScript API that may result in unexpected behavior or require specific coding patterns to achieve the desired outcome. If you encounter an issue that belongs in this list, please let us know by using the feedback form at the bottom of the article.

> ## Common and Outlook APIs are not promise-based

Reviewer: ```suggestion
## Common APIs and Outlook APIs are not promise-based
```

> The [Common APIs](/javascript/api/office) (those that are not tied to a particular host) and [Outlook APIs](/javascript/api/outlook) use a callback-based programming model. Interacting with the underlying Office document requires an asynchronous read or write call that includes a callback to be ran when the operation completes. For an example of this pattern, see [AppointmentCompose.getAttachmentAsync](/javascript/api/office/office.file?view=common-js#getsliceasync-sliceindex--callback-).

Reviewer: ```suggestion
The [Common APIs](/javascript/api/office) (those that are not tied to a particular Office host) and [Outlook APIs](/javascript/api/outlook) use a callback-based programming model. Interacting with the underlying Office document requires an asynchronous read or write call that specifies a callback to be ran when the operation completes. For an example of this pattern, see [AppointmentCompose.getAttachmentAsync](/javascript/api/office/office.file?view=common-js#getsliceasync-sliceindex--callback-).
```

> The [Common APIs](/javascript/api/office) (those that are not tied to a particular host) and [Outlook APIs](/javascript/api/outlook) use a callback-based programming model. Interacting with the underlying Office document requires an asynchronous read or write call that includes a callback to be ran when the operation completes. For an example of this pattern, see [AppointmentCompose.getAttachmentAsync](/javascript/api/office/office.file?view=common-js#getsliceasync-sliceindex--callback-).

Reviewer: The URL target (bookmark `#getsliceasync-sliceindex--callback-`) specified at the end of this paragraph for the link `AppointmentCompose.getAttachmentAsync` doesn't point to the right place in the reference docs.

> These Common and Outlook methods do not return [Promises](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise). As such, you cannot use [await](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Operators/await) to pause the execution until the async operation completes. If you need `await` behavior, you can wrap the method call in an explicitly created Promise.

Reviewer: ```suggestion
These Common API methods and Outlook API methods do not return [Promises](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise). Therefore, you cannot use [await](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Operators/await) to pause the execution until the asynchronous operation completes. If you need `await` behavior, you can wrap the method call in an explicitly created Promise.
```

> These Common and Outlook methods do not return [Promises](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise). As such, you cannot use [await](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Operators/await) to pause the execution until the async operation completes. If you need `await` behavior, you can wrap the method call in an explicitly created Promise.

Reviewer: Would it be helpful to include a code sample here that demonstrates this approach?

Author: Probably. I'll craft and test one.

- Read-only property: Subproperties can be set through navigation.
- Writable property: Subproperties must be set with a JSON struct (and cannot be set through navigation).

> ## Excel Range limits

Reviewer: I'm wondering if (for discoverability), it'd be better to add this information to the [Ranges (advanced)](https://docs.microsoft.com/en-us/office/dev/add-ins/excel/excel-add-ins-ranges-advanced) article, instead of here in this coding issues article?  Or, perhaps define the contents of the section in an include file, and include it both here in the coding issues article and also in the Excel Ranges (advanced) article? Thoughts?

Author: There's a section in the Excel Core Concepts doc that mentions the 5MB limit. I'll replace that with a link to this section. I'll also add a link to the new coding issues page from the **See Also** list.

> Excel developers need to be aware of two data size limitations:

Reviewer: ```suggestion
If you're building an Excel add-in that uses ranges, be aware of the following size limitations:
```

- Excel on the web has a payload size limit for requests and responses of 5MB. `RichAPI.Error` will be thrown if that limit is exceeded.
- A range is limited to five million cells for set operations.

> If you expect user input to exceed those amounts, be sure to check the data and split the ranges into multiple objects. You'll also need separate `context.sync()` calls to avoid overwhelming the Excel cloud service.

Reviewer: ```suggestion
If you expect user input to exceed these limits, be sure to check the data and split the ranges into multiple objects. You'll also need to submit multiple `context.sync()` calls to optimize performance.
```

Author: I'm taking the first two changes. To the third point, the issue isn't performance. They'll need to use multiple syncs to avoid the size limit error. I'll try to highlight that point better.
